### PR TITLE
feat: add kakarotRpc struct to better manager our RPC implementation

### PR DESCRIPTION
**Pull request by bajpai244**
In this proposed solution, every time we want to add an API, we put it in the API directory, then we implement a server for the API, and then we can update our KakarotRPC struct.

KakarotRPC struct can then expose a single `RpcModule`, which can run all these methods from these various traits on the same server and port, we are able to separate concerns, at the same time during runtime able to run them all under a single server.

solves #291

### CODE REVIEW

File: crates/core/src/client/config.rs [hunk 1]
```diff
@@ -48,6 +48,16 @@ impl StarknetConfig {
     }
 }
 
+impl Clone for StarknetConfig {
```
Comment (Eikix) [inline crates/core/src/client/config.rs:51]: What is the purpose of this manual implementation of clone?

- I suspect you did this because you can't just derive Clone for StarknetConfig?
- Is there a workaround for this? Namely, use references when needed rather than clones or functions that consume the ownership?

File: crates/eth-rpc/src/rpc.rs [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+use crate::api::eth_api::EthApiServer;
+use crate::api::kakarot_custom_api::KakarotCustomApiServer;
+use crate::servers::eth_rpc;
+use crate::servers::eth_rpc::KakarotEthRpc;
+use crate::servers::kakarot_custom_rpc::{self, KakarotCustomRpc};
+
+pub struct KakarotRpc {
+    eth_rpc_module: RpcModule<KakarotEthRpc<JsonRpcClient<HttpTransport>>>,
+    kakarot_custom_rpc_module: RpcModule<KakarotCustomRpc<JsonRpcClient<HttpTransport>>>,
```
Comment (Eikix) [inline crates/eth-rpc/src/rpc.rs:15]: Same here, especially if they use the same type (`JsonRpcClient<HttpTransport>`), it could be passed as a generic to `KakarotRpc`

File: crates/eth-rpc/src/rpc.rs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+            eth_rpc::KakarotEthRpc::new(Box::new(KakarotRpc::get_kakarot_client(starknet_config.clone()))).into_rpc();
+        let kakarot_custom_rpc_module =
+            kakarot_custom_rpc::KakarotCustomRpc::new(Box::new(KakarotRpc::get_kakarot_client(starknet_config)))
+                .into_rpc();
+
+        Self { eth_rpc_module, kakarot_custom_rpc_module }
+    }
+
+    pub fn get_rpc_module(&self) -> RpcModule<()> {
+        let mut rpc_module = RpcModule::new(());
```
Comment (Eikix) [inline crates/eth-rpc/src/rpc.rs:30]: is there a slightly cleaner way to do this merging?

File: crates/eth-rpc/src/rpc.rs [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+        rpc_module.merge(self.eth_rpc_module.clone()).unwrap();
```
Comment (Eikix) [inline crates/eth-rpc/src/rpc.rs:32]: why do we unwrap here? is there a way around it?

File: crates/eth-rpc/src/rpc.rs [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+        let eth_rpc_module =
+            eth_rpc::KakarotEthRpc::new(Box::new(KakarotRpc::get_kakarot_client(starknet_config.clone()))).into_rpc();
+        let kakarot_custom_rpc_module =
+            kakarot_custom_rpc::KakarotCustomRpc::new(Box::new(KakarotRpc::get_kakarot_client(starknet_config)))
+                .into_rpc();
+
+        Self { eth_rpc_module, kakarot_custom_rpc_module }
+    }
+
+    pub fn get_rpc_module(&self) -> RpcModule<()> {
```
Comment (Eikix) [inline crates/eth-rpc/src/rpc.rs:29]: can we add rustdoc comments
akin to what @jobez did in his recent PRs:)?

just explain the general flow of things

File: crates/eth-rpc/src/rpc.rs [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        let mut rpc_module = RpcModule::new(());
+
+        rpc_module.merge(self.eth_rpc_module.clone()).unwrap();
+        rpc_module.merge(self.kakarot_custom_rpc_module.clone()).unwrap();
+
+        rpc_module
+    }
+
+    fn get_kakarot_client(starknet_config: StarknetConfig) -> KakarotClient<JsonRpcClient<HttpTransport>> {
+        let provider = JsonRpcClientBuilder::with_http(&starknet_config).unwrap().build();
```
Comment (Eikix) [inline crates/eth-rpc/src/rpc.rs:39]: why unwrap? was this already there?

File: crates/eth-rpc/src/servers/kakarot_custom_rpc.rs [hunk 7]
```diff
@@ -0,0 +1,27 @@
+use jsonrpsee::core::{async_trait, RpcResult as Result};
+use kakarot_rpc_core::client::api::KakarotEthApi;
+use kakarot_rpc_core::models::balance::TokenBalances;
+use reth_primitives::Address;
+use starknet::providers::Provider;
+
+use crate::api::kakarot_custom_api::KakarotCustomApiServer;
+
+/// The RPC module for the Ethereum protocol required by Kakarot.
+pub struct KakarotCustomRpc<P: Provider + Send + Sync> {
```
Comment (Eikix) [inline crates/eth-rpc/src/servers/kakarot_custom_rpc.rs:10]: I think I got this API from alchemy
https://docs.alchemy.com/reference/ethereum-api-endpoints

let's rename it to give them many thanks

File: crates/eth-rpc/src/rpc.rs [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+use crate::servers::eth_rpc::KakarotEthRpc;
+use crate::servers::kakarot_custom_rpc::{self, KakarotCustomRpc};
+
+pub struct KakarotRpc {
+    eth_rpc_module: RpcModule<KakarotEthRpc<JsonRpcClient<HttpTransport>>>,
+    kakarot_custom_rpc_module: RpcModule<KakarotCustomRpc<JsonRpcClient<HttpTransport>>>,
+}
+
+impl KakarotRpc {
+    pub fn new(starknet_config: StarknetConfig) -> Self {
```
Comment (Eikix) [inline crates/eth-rpc/src/rpc.rs:19]: a problem I see here is that we're using 2 different `KakarotClient` instances, one for each API, but they're the same client! 
It'll end up being quite inefficient

Possible workarounds:
Having Arc clones of KakarotClient passed around?

You'll have to investigate if that doesn't work out -> our main philosophical challenge here is: each api will probably point to core crate, thus kakarot client, how can we not duplicate kakarot client instances for each separate api

**Review by Eikix — CHANGES_REQUESTED** [review]
great PR! going to give us lots of freedom to add new APIs to our json rpc server

let me know if comments are clear

**Comment by bajpai244** [comment]
> great PR! going to give us lots of freedom to add new APIs to our json rpc server
> 
> let me know if comments are clear

@Eikix thanks for the review, working on these comments by resolving one at a time.

I would also consider to watch for PRs we are going to merge before this, to see if there can be big rebase issues.

On it ⚡

File: crates/eth-rpc/src/rpc.rs [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    pub fn get_rpc_module(&self) -> RpcModule<()> {
+        let mut rpc_module = RpcModule::new(());
+
+        rpc_module.merge(self.eth_rpc_module.clone()).unwrap();
+        rpc_module.merge(self.kakarot_custom_rpc_module.clone()).unwrap();
+
+        rpc_module
+    }
+
+    fn get_kakarot_client(starknet_config: StarknetConfig) -> KakarotClient<JsonRpcClient<HttpTransport>> {
```
[the remark this answers is not in the document]
Comment (bajpai244) [reply crates/eth-rpc/src/rpc.rs]: Done.

File: crates/eth-rpc/src/rpc.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (bajpai244) [reply crates/eth-rpc/src/rpc.rs]: Done.

File: crates/eth-rpc/src/servers/kakarot_custom_rpc.rs (hunk 7 above)
Comment (bajpai244) [reply crates/eth-rpc/src/servers/kakarot_custom_rpc.rs:10]: Makes sense. Resolved.

Also, then it means we are not completely implementing the Token Api, we just have one method, should I make an issue for this?

image

File: crates/core/src/client/config.rs (hunk 1 above)
Comment (ClementWalter) [reply crates/core/src/client/config.rs:51]: Actually I was able to compile

```
#[derive(Default, Clone)]
/// Configuration for the Starknet RPC client.
pub struct StarknetConfig {
    /// Rpc url.
    pub url: String,
    /// Kakarot contract address.
    pub kakarot_address: FieldElement,
    /// Proxy account class hash.
    pub proxy_account_class_hash: FieldElement,
}
```

File: crates/eth-rpc/src/api/kakarot_custom_api.rs [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+
+#[rpc(server)]
+#[async_trait]
+pub trait KakarotCustomApi {
+    #[method(name = "kakarot_getTokenBalances")]
+    async fn token_balances(
+        &self,
+        address: Address,
+        contract_addresses: Vec<Address>,
+    ) -> jsonrpsee::core::RpcResult<TokenBalances>;
```
Comment (ClementWalter) [inline crates/eth-rpc/src/api/kakarot_custom_api.rs:13]: can we keep the same semantic as in the other trait, ie import 

```
use jsonrpsee::core::RpcResult as Result;
```

and return

```
Result<TokenBalances>
```

File: crates/eth-rpc/src/servers/kakarot_custom_rpc.rs (hunk 7 above)
Comment (ClementWalter) [reply crates/eth-rpc/src/servers/kakarot_custom_rpc.rs:10]: imho these `alchemy_` endpoints are very low priority for us. It is was good catch to realize that we didn't serve what we thought we were serving, and we definitely could define one trait per prefix even in the "standard" et json rpc spec (`eth_`, `net_`, `web3_`). This `alchemy_` RPC is very low priority for us, so why not an issue, but with "nice to have", and "low priority" tags!

File: crates/eth-rpc/src/api/kakarot_custom_api.rs [hunk 11]
```diff
@@ -0,0 +1,14 @@
+use jsonrpsee::proc_macros::rpc;
+use kakarot_rpc_core::models::balance::TokenBalances;
+use reth_primitives::Address;
+
+#[rpc(server)]
+#[async_trait]
+pub trait KakarotCustomApi {
```
Comment (ClementWalter) [inline crates/eth-rpc/src/api/kakarot_custom_api.rs:7]: so this would be `AlchemyApi`, and file renamed accordingly

File: crates/eth-rpc/src/api/kakarot_custom_api.rs [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+    #[method(name = "kakarot_getTokenBalances")]
```
Comment (ClementWalter) [inline crates/eth-rpc/src/api/kakarot_custom_api.rs:8]: isn't there a way (I think there is) to automatically prepend the `alchemy_` (here `kakarot_`) to all `#[methods()]` (maybe create an issue and let's update this for the eth api as well)

File: crates/eth-rpc/src/rpc.rs [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+use starknet::providers::JsonRpcClient;
+
+use crate::api::eth_api::EthApiServer;
+use crate::api::kakarot_custom_api::KakarotCustomApiServer;
+use crate::servers::eth_rpc;
+use crate::servers::eth_rpc::KakarotEthRpc;
+use crate::servers::kakarot_custom_rpc::{self, KakarotCustomRpc};
+
+pub struct KakarotRpc {
+    eth_rpc_module: RpcModule<KakarotEthRpc<JsonRpcClient<HttpTransport>>>,
```
[the remark this answers is not in the document]
Comment (ClementWalter) [reply crates/eth-rpc/src/rpc.rs:14]: shouldn't it be `Provider` instead?

**Review by ClementWalter — COMMENTED** [review]
This PR goes to me into the right direction. I'm still confused by our

```
impl<P: Provider + Send + Sync> KakarotEthApi<P> for KakarotClient<P>
```

but this for clarifies already the structure of the RPC

File: crates/eth-rpc/src/rpc.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (bajpai244) [reply crates/eth-rpc/src/rpc.rs]: @ClementWalter agreed, @Eikix I have pushed new changes which makes the whole struct generic over the provider.

image

File: crates/eth-rpc/src/rpc.rs (hunk 2 above)
Comment (bajpai244) [reply crates/eth-rpc/src/rpc.rs]: This is also now generic over `Provider`.

image

File: crates/eth-rpc/src/rpc.rs (hunk 6 above)
Comment (bajpai244) [reply crates/eth-rpc/src/rpc.rs]: We no more need this function with the new changes I have pushed.

File: crates/eth-rpc/src/servers/kakarot_custom_rpc.rs (hunk 7 above)
Comment (bajpai244) [reply crates/eth-rpc/src/servers/kakarot_custom_rpc.rs]: @ClementWalter , noted, whenI make the issue, I will make sure to use these tags.

File: crates/eth-rpc/src/rpc.rs (hunk 8 above)
Comment (bajpai244) [reply crates/eth-rpc/src/rpc.rs]: Resolved with the new code changes I pushed.

image

File: crates/core/src/client/config.rs (hunk 1 above)
Comment (bajpai244) [reply crates/core/src/client/config.rs]: So, we no more need this, also yeah I could have simply derived, but I guess my Rust noobness made me implement the trait manually! 

TLDR;

We no longer need clone.

File: crates/eth-rpc/src/rpc.rs (hunk 5 above)
Comment (jobez) [reply crates/eth-rpc/src/rpc.rs]: gpt does a decent job at this, but give yourself time to read through it yourself before you push

File: crates/eth-rpc/src/servers/kakarot_custom_rpc.rs (hunk 7 above)
Comment (bajpai244) [reply crates/eth-rpc/src/servers/kakarot_custom_rpc.rs]: It has been renamed accordingly -> 
image

**Comment by bajpai244** [comment]
> > What is the purpose of this manual implementation of clone?
> 
> also wondering when we use borrow and when we take ownership of the args in out functions

We no longer need it.

File: crates/eth-rpc/src/api/kakarot_custom_api.rs (hunk 10 above)
Comment (bajpai244) [reply crates/eth-rpc/src/api/kakarot_custom_api.rs]: Done.

File: crates/eth-rpc/src/api/kakarot_custom_api.rs (hunk 11 above)
Comment (bajpai244) [reply crates/eth-rpc/src/api/kakarot_custom_api.rs]: Yes, we can create namespaces and do that, I am planning a next PR after it for that,

because:
- We would have to do that in EthApi as well
-  I can see net_version doesn't start with "eth"
-  which means we probably would have to refactor a new trait for it, and I think in that case I would prefer to do it in another PR, lemme know what you think. 
image

File: crates/eth-rpc/src/api/kakarot_custom_api.rs (hunk 12 above)
Comment (bajpai244) [reply crates/eth-rpc/src/api/kakarot_custom_api.rs]: Yes, we can create namespaces and do that, I am planning a next PR after it for that,

because:
- We would have to do that in EthApi as well
-  I can see net_version doesn't start with "eth"
-  which means we probably would have to refactor a new trait for it, and I think in that case I would prefer to do it in another PR, lemme know what you think. 
image

File: crates/eth-rpc/src/rpc.rs (hunk 4 above)
Comment (bajpai244) [reply crates/eth-rpc/src/rpc.rs]: I am returning a Result now { I was hacking fast on the PR, and then pulled it out from draft on daily without remembering about it!  }

image

File: crates/eth-rpc/src/main.rs [hunk 14]
```diff
@@ -17,7 +21,11 @@ async fn main() -> Result<()> {
     let starknet_config = StarknetConfig::from_env()?;
     let rpc_config = RPCConfig::from_env()?;
 
-    let kakarot_rpc = KakarotRpc::new(starknet_config);
+    let provider = JsonRpcClientBuilder::with_http(&starknet_config).unwrap().build();
```
Comment (Eikix) [inline crates/eth-rpc/src/main.rs:24]: no need to unwrap here I think

File: crates/eth-rpc/src/main.rs [hunk 15]
```diff
@@ -17,7 +21,11 @@ async fn main() -> Result<()> {
...
+    let kakarot_client: Arc<Box<dyn KakarotEthApi<_>>> =
+        Arc::new(Box::new(KakarotClient::new(starknet_config, provider)));
```
Comment (Eikix) [inline crates/eth-rpc/src/main.rs:26]: quick sanity check: is an Arc of a Box considered "ok" in Rust?

File: crates/eth-rpc/src/rpc.rs [hunk 16]
```diff
@@ -1,41 +1,34 @@
...
+    token_rpc_module: RpcModule<TokenRpc<P>>,
 }
 
-impl KakarotRpc {
-    pub fn new(starknet_config: StarknetConfig) -> Self {
-        let eth_rpc_module =
-            eth_rpc::KakarotEthRpc::new(Box::new(KakarotRpc::kakarot_client(starknet_config.clone()))).into_rpc();
-        let kakarot_custom_rpc_module =
-            kakarot_custom_rpc::KakarotCustomRpc::new(Box::new(KakarotRpc::kakarot_client(starknet_config))).into_rpc();
+impl<P: Provider + Send + Sync + 'static> KakarotRpc<P> {
```
Comment (Eikix) [inline crates/eth-rpc/src/rpc.rs:19]: nice!

File: crates/eth-rpc/src/rpc.rs (hunk 5 above)
Comment (Eikix) [inline crates/eth-rpc/src/rpc.rs]: get_rpc_module -> rpc_module ("get" idiomatic rust)

and let's add rustdoc

File: crates/eth-rpc/src/rpc.rs (hunk 6 above)
Comment (Eikix) [reply crates/eth-rpc/src/rpc.rs]: what do you mean?

File: crates/eth-rpc/src/main.rs (hunk 14 above)
Comment (Eikix) [reply crates/eth-rpc/src/main.rs:24]: ah well this was done before^^ interesting. 
i think we should utilize `eyre` to exit gracefully rather than unwrap

Comment (Eikix) [reply crates/eth-rpc/src/main.rs:24]: if not, we can also use "expect"

Comment (Eikix) [reply crates/eth-rpc/src/main.rs:24]: can be done in another issue, let's just create the issue and tackle this in another PR
great job

**Review by Eikix — APPROVED** [review]
lgtm! gg @bajpai244

File: crates/eth-rpc/src/rpc.rs (hunk 3 above)
Comment (bajpai244) [reply crates/eth-rpc/src/rpc.rs]: Did this refactor which is cleaner, lemme know what your thoughts are.

Took some [inspiration](https://github.com/paradigmxyz/reth/blob/dbafe23cce45917a43f6640d71a0216fe3268428/crates/rpc/rpc-builder/src/lib.rs#L632) from reth, tried to keep our implementation as simple as possible.

image

File: crates/eth-rpc/src/rpc.rs [hunk 17]
```diff
@@ -1,34 +1,49 @@
...
+#[derive(Debug, Clone, Copy, Eq, PartialEq, Hash)]
+pub enum KakarotRpcModule {
+    Eth,
+    Alchemy,
 }
 
-impl<P: Provider + Send + Sync + 'static> KakarotRpc<P> {
+pub struct KakarotRpcModuleBuilder<P: Provider + Send + Sync + 'static> {
+    modules: HashMap<KakarotRpcModule, Methods>,
+    _phantom: PhantomData<P>,
```
Comment (Eikix) [inline crates/eth-rpc/src/rpc.rs:24]: why do we need phantom data?

File: crates/eth-rpc/src/rpc.rs [hunk 18]
```diff
@@ -1,34 +1,49 @@
...
+
+        modules.insert(KakarotRpcModule::Eth, eth_rpc_module.into());
+        modules.insert(KakarotRpcModule::Alchemy, alchemy_rpc_module.into());
 
-        Self { eth_rpc_module, token_rpc_module }
+        Self { modules, _phantom: PhantomData }
     }
 
-    pub fn rpc_module(&self) -> RpcModule<()> {
+    pub fn rpc_module(&self) -> Result<RpcModule<()>, Error> {
```
Comment (Eikix) [inline crates/eth-rpc/src/rpc.rs:40]: i guess following the builder pattern, this fn should be name `build` or `build_rpc_module`

File: crates/eth-rpc/src/lib.rs [hunk 19]
```diff
@@ -25,7 +25,7 @@ pub enum RpcError {
 ///
 /// Will return `Err` if an error occurs when running the `ServerBuilder` start fails.
 pub async fn run_server<P: Provider + Send + Sync + 'static>(
```
Comment (Eikix) [inline crates/eth-rpc/src/lib.rs:27]: I believe we should keep the same signature for run_server, and trigger the build beforehand *** C.F below comment

File: crates/eth-rpc/src/main.rs [hunk 20]
```diff
@@ -25,7 +25,7 @@ async fn main() -> Result<()> {
     let kakarot_client: Arc<Box<dyn KakarotEthApi<_>>> =
         Arc::new(Box::new(KakarotClient::new(starknet_config, provider)));
 
-    let kakarot_rpc = KakarotRpc::new(kakarot_client);
+    let kakarot_rpc = KakarotRpcModuleBuilder::new(kakarot_client);
```
Comment (Eikix) [inline crates/eth-rpc/src/main.rs:28]: > I believe we should keep the same signature for run_server, and trigger the build beforehand *** C.F below comment

We should trigger the build right here

File: crates/eth-rpc/src/lib.rs (hunk 19 above)
Comment (bajpai244) [reply crates/eth-rpc/src/lib.rs]: Done.

File: crates/eth-rpc/src/rpc.rs (hunk 18 above)
Comment (bajpai244) [reply crates/eth-rpc/src/rpc.rs]: makes sense, refactored to `build_rpc_module`

File: crates/eth-rpc/src/rpc.rs (hunk 17 above)
Comment (bajpai244) [reply crates/eth-rpc/src/rpc.rs:24]: Without it the compiler will throw error for type parameter `P`

image

File: crates/eth-rpc/src/main.rs (hunk 14 above)
Comment (bajpai244) [reply crates/eth-rpc/src/main.rs:24]: Gotcha! So that makes two issues we have to create:
- Implement all Alchemy endpoints { low priority }
- Refactor to avoid unwrapping here.

After the merge if you want I can create them, or if you want to then that is also fine.

**Comment by Eikix** [comment]
you got conflicts, rip was close to merge first

File: crates/eth-rpc/src/main.rs (hunk 14 above)
Comment (Eikix) [reply crates/eth-rpc/src/main.rs:24]: Yes! you can create them sounds good

**Review by Eikix — APPROVED** [review]
great job!
